extensions/ql-vscode/src/data-extensions-editor/extension-pack-picker.ts

```diff
@@ -303,12 +308,145 @@ async function pickNewExtensionPack(
   }
 
   const name = matches.groups.name;
-  const packPath = join(workspaceFolder.path, name);
+  const packPath = join(workspaceFolder.uri.fsPath, name);
 
   if (await pathExists(packPath)) {
     return undefined;
   }
 
+  return writeExtensionPack(packPath, packName, databaseItem.language);
+}
+
+async function autoCreateExtensionPack(
+  name: string,
+  language: string,
+  extensionPacksInfo: QlpacksInfo,
+  logger: NotificationLogger,
+): Promise<ExtensionPack | undefined> {
+  const workspaceFolder = await autoPickWorkspaceFolder(language);
+  if (!workspaceFolder) {
+    return undefined;
+  }
+
+  const packName = autoNameExtensionPack(name, language);
+  if (!packName) {
+    void showAndLogErrorMessage(
+      logger,
+      `Could not automatically name extension pack for database ${name}`,
+    );
+
+    return undefined;
+  }
+
+  const existingExtensionPackPaths = extensionPacksInfo[packName];
+  if (existingExtensionPackPaths?.length === 1) {
+    let extensionPack: ExtensionPack;
+    try {
+      extensionPack = await readExtensionPack(existingExtensionPackPaths[0]);
+    } catch (e: unknown) {
+      void showAndLogErrorMessage(
+        logger,
+        `Could not read extension pack ${packName}`,
+        {
+          fullMessage: `Could not read extension pack ${packName} at ${
+            existingExtensionPackPaths[0]
+          }: ${getErrorMessage(e)}`,
+        },
+      );
+
+      return undefined;
+    }
+
+    return extensionPack;
+  } else if (existingExtensionPackPaths?.length > 1) {
+    void showAndLogErrorMessage(
+      logger,
+      `Extension pack ${packName} resolves to multiple paths`,
+      {
+        fullMessage: `Extension pack ${packName} resolves to multiple paths: ${existingExtensionPackPaths.join(
+          ", ",
+        )}`,
+      },
+    );
+
+    return undefined;
+  }
+
+  const matches = packNameRegex.exec(packName);
+  if (!matches?.groups) {
+    void showAndLogErrorMessage(
+      logger,
+      `Extension pack ${packName} does not have a valid name`,
+    );
+
+    return undefined;
+  }
+
+  const unscopedName = matches.groups.name;
+  const packPath = join(workspaceFolder.uri.fsPath, unscopedName);
+
+  if (await pathExists(packPath)) {
+    void showAndLogErrorMessage(
+      logger,
+      `Directory ${packPath} already exists for extension pack ${packName}`,
+    );
+
+    return undefined;
+  }
+
+  return writeExtensionPack(packPath, packName, language);
+}
+
+async function autoPickWorkspaceFolder(
+  language: string,
+): Promise<WorkspaceFolder | undefined> {
+  const workspaceFolders = getOnDiskWorkspaceFoldersObjects();
+
+  if (workspaceFolders.length === 1) {
+    return workspaceFolders[0];
+  }
+  const starterWorkspaceFolderForLanguage = workspaceFolders.find(
+    (folder) => folder.name === `codeql-custom-queries-${language}`,
```

Comment:
Is it ok to have a pack inside a pack?

Reply:
Yes, that works. I agree that it isn't ideal, but I'm not sure where to create the extension packs otherwise. There isn't a workspace folder in the starter workspace that doesn't have a query pack.

Reply:
I can't think of an immediate thing that would break, but this sounds potentially very confusing, since query packs are not usually meant to contain data extensions themselves, or nested packs (except when compiled, in which case they include their library pack dependencies).
I would rather put an empty placeholder folder now into the starter workspace, and then populate your generated extension packs into that folder.

Also where will extension packs go for repos that are not the starter workspace?

Reply:
Possibly, we can add `.github/codeql/extensions` as a workspace folder (possibly empty) and put the extension packs there. This folder is the location where internal extensions are kept. Even though it's not directly useful in the starter repo, this folder is recognized for code scanning runs.

And, I agree with Aditya. I don't see any technical problem of nesting packs in this case, but it is confusing.

Reply:
I've now changed it to put all extension packs in `.github/codeql/extensions`, unless the extension pack already exists. This should remove the nested extension packs and make it clearer where extension packs will go in almost all cases. It will now automatically add a new workspace folder for `.github/codeql/extensions`, unless the user is in a single-root workspace. In that case, it will be added in that workspace folder directly in the `.github/codeql/extensions` directory.

Previously, it would try to automatically determine the workspace folder by checking for the `-<language>` suffix, and otherwise it would just ask the user.